github/repos_contents.go

```diff
@@ -267,6 +267,12 @@ func (s *RepositoriesService) GetArchiveLink(owner, repo string, archiveformat a
 	} else {
 		resp, err = s.client.client.Transport.RoundTrip(req)
 	}
+	defer func() {
+		// Make sure to close the response body if we have one
+		if resp != nil {
+			resp.Body.Close()
+		}
+	}()
 	if err != nil || resp.StatusCode != http.StatusFound {
 		return nil, newResponse(resp), err
 	}
```

Comment:
The above code will work, but I have a suggestion. The canonical example of doing an HTTP request looks like this:

```Go
resp, err := http.Get("http://example.com/")
if err != nil {
	// handle error
}
defer resp.Body.Close()
body, err := ioutil.ReadAll(resp.Body)
// ...
```

Notice the error is checked first, and ensuring `resp.Body` will be closed is after. I think it'd be cleaner to rewrite it in a similar way.

First check if `err != nil` from `RoundTrip`, handle that, then handle success case. It would require breaking up the `if err != nil || resp.StatusCode != http.StatusFound {` check into two, but that seems very appropriate. So it could look something like this:

```Go
var resp *http.Response
// Use http.DefaultTransport if no custom Transport is configured
if s.client.client.Transport == nil {
	resp, err = http.DefaultTransport.RoundTrip(req)
} else {
	resp, err = s.client.client.Transport.RoundTrip(req)
}
if err != nil {
	return nil, nil, err
}
resp.Body.Close() // Close the body right away, it's unused.
if resp.StatusCode != http.StatusFound {
	return nil, newResponse(resp), err
}
```

`defer` in front of `resp.Body.Close()` isn't even needed because we can just close it right away instead of when returning. It's never used.

What do you think?